File: tools/mmd-dumper/scripts/package-native.mjs
#!/usr/bin/env node
import { copyFile, mkdir, readFile, writeFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import { basename, dirname, resolve } from "node:path";

const root = resolve(import.meta.dirname, "..");
const outDir = resolve(root, "out", "mmd-oracle-dumper-package");
const artifacts = [
  {
    name: "mmd_oracle_dumper.dll",
    source: resolve(root, "out", "native-dll-smoke", "mmd_oracle_dumper.dll"),
    required: true,
  },
  {
    name: "MSIMG32.dll",
    source: resolve(root, "out", "native-proxy-smoke", "MSIMG32.dll"),
    required: true,
  },
  {
    name: "d3d9.dll",
    source: resolve(root, "out", "native-d3d9-smoke", "d3d9.dll"),
    required: false,
  },
  {
    name: "Plugin/mmd_oracle_plugin.dll",
    source: resolve(root, "out", "native-mmdplugin-smoke", "mmd_oracle_plugin.dll"),
    required: false,
  },
  {
    name: "README.md",
    source: resolve(root, "native", "README.md"),
    required: true,
  },
  {
    name: "oracle-v1.schema.json",
    source: resolve(root, "schema", "oracle-v1.schema.json"),
    required: true,
  },
];

await mkdir(outDir, { recursive: true });

const copied = [];
for (const artifact of artifacts) {
  if (!existsSync(artifact.source)) {
    if (artifact.required) {
      throw new Error(`Required artifact is missing: ${artifact.source}. Run pnpm -C MMDDumper native:test first.`);
    }
    continue;
  }
  const destination = resolve(outDir, artifact.name);
  await mkdir(dirname(destination), { recursive: true });
  await copyFile(artifact.source, destination);
  copied.push({ name: artifact.name, source: artifact.source, destination });
}

const exampleDumpPath = resolve(root, "fixtures", "sample-basic", "oracle.actual.jsonl").replaceAll("\\", "\\\\");
const installText = `# MMD Oracle Dumper local package

Preferred MMDPlugin install:

- Keep the MMDPlugin-provided \`d3d9.dll\` next to \`MikuMikuDance.exe\`.
- Copy \`Plugin/mmd_oracle_plugin.dll\` into the local MMD \`Plugin/\` folder.

Legacy proxy fallback:

- mmd_oracle_dumper.dll
- MSIMG32.dll
- d3d9.dll, only when intentionally using the old d3d9 proxy trigger

Optional first-load smoke environment variables:

\`\`\`powershell
$env:MMD_ORACLE_DUMP_PATH = "${exampleDumpPath}"
$env:MMD_ORACLE_DUMP_ON_PROXY_LOAD = "1"
$env:MMD_ORACLE_DUMP_ON_D3D9 = "1"
$env:MMD_ORACLE_DUMP_ON_MMDPLUGIN = "1"
\`\`\`

This package is a local oracle recorder prototype. It is not a generic injector and does not install anything globally.
`;

await writeFile(resolve(outDir, "INSTALL.md"), installText, "utf8");

const schema = JSON.parse(await readFile(resolve(root, "schema", "oracle-v1.schema.json"), "utf8"));
const manifest = {
  name: "mmd-oracle-dumper",
  schema: schema.$id,
  mmdVersion: "9.32-x64",
  files: copied.map((entry) => ({
    name: entry.name,
    destinationHint: entry.name === "oracle-v1.schema.json" || entry.name === "README.md" ? "documentation" : "next to MikuMikuDance.exe",
  })),
  safety: {
    genericInjector: false,
    globalInstall: false,
    networkAccess: false,
    targetProcess: "MikuMikuDance.exe",
  },
};

await writeFile(resolve(outDir, "package-manifest.json"), `${JSON.stringify(manifest, null, 2)}\n`, "utf8");

console.log(
  JSON.stringify(
    {
      ok: true,
      outDir,
      files: copied.map((entry) => basename(entry.destination)),
    },
    null,
    2,
  ),
);

File: MMDDumper/scripts/package_native.py
#!/usr/bin/env python3
"""Package locally-built MMDDumper native artifacts without Node.js."""

from __future__ import annotations

import json
import shutil
from pathlib import Path


def main() -> int:
    root = Path(__file__).resolve().parents[1]
    out_dir = root / "out" / "mmd-oracle-dumper-package"
    artifacts = (
        ("mmd_oracle_dumper.dll", _built_artifact(root, "native-dll-smoke", "mmd_oracle_dumper.dll"), True),
        ("MSIMG32.dll", _built_artifact(root, "native-proxy-smoke", "MSIMG32.dll"), True),
        ("d3d9.dll", _built_artifact(root, "native-d3d9-smoke", "d3d9.dll"), False),
        ("Plugin/mmd_oracle_plugin.dll", _built_artifact(root, "native-mmdplugin-smoke", "mmd_oracle_plugin.dll"), False),
        ("oracle-v1.schema.json", root / "schema" / "oracle-v1.schema.json", True),
    )
    out_dir.mkdir(parents=True, exist_ok=True)
    copied: list[dict[str, str]] = []
    for relative, source, required in artifacts:
        if not source.is_file():
            if required:
                raise FileNotFoundError(f"required artifact is missing: {source}; run native_smoke.py first")
            continue
        destination = out_dir / relative
        destination.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(source, destination)
        copied.append({"name": relative, "source": str(source), "destination": str(destination)})

    readme = out_dir / "README.md"
    readme.write_text(
        "# MMDDumper native package\n\n"
        "Copy the MMDPlugin DLL into the local MMD `Plugin/` directory.\n"
        "The legacy proxy DLLs are retained for local diagnostics only.\n"
        "This package is repository-local and never installs files globally.\n",
        encoding="utf-8",
    )
    copied.append({"name": "README.md", "source": str(readme), "destination": str(readme)})
    schema = json.loads((root / "schema" / "oracle-v1.schema.json").read_text(encoding="utf-8"))
    manifest = {
        "name": "mmd-oracle-dumper",
        "schema": schema["$id"],
        "mmdVersion": "9.32-x64",
        "files": [{"name": item["name"], "destinationHint": "documentation" if item["name"].endswith((".json", ".md")) else "next to MikuMikuDance.exe"} for item in copied],
        "safety": {"genericInjector": False, "globalInstall": False, "networkAccess": False, "targetProcess": "MikuMikuDance.exe"},
    }
    (out_dir / "package-manifest.json").write_text(json.dumps(manifest, ensure_ascii=True, indent=2) + "\n", encoding="utf-8")
    print(json.dumps({"ok": True, "outDir": str(out_dir), "files": [item["name"] for item in copied]}, ensure_ascii=True, indent=2))
    return 0


def _built_artifact(root: Path, build: str, name: str) -> Path:
    directory = root / "out" / build
    candidates = (directory / name, directory / "Release" / name)
    return next((path for path in candidates if path.is_file()), candidates[0])


if __name__ == "__main__":
    raise SystemExit(main())
